### ISSUE TITLE
Datatypes with arrows

### ISSUE THREAD

**Issue by sweirich** [goldfirere/singletons#510]
This is what I want to do:

```
$(singletons [d| 
     newtype Dyn = D (Dyn -> Dyn)
 |])
```

I'm getting this error (which I understand, but I don't know how to work around.)

```
Couldn't match kind ‘TyFun Dyn Dyn’ with ‘Dyn’
 Expected kind ‘Dyn -> Dyn’,
   but ‘a6989586621679680017’ has kind ‘Dyn ~> Dyn’
• In the first argument of ‘D’, namely ‘a6989586621679680017’
 In the type ‘D a6989586621679680017’
 In the type family declaration for ‘DSym1’
```

Richard says try this:
```
newtype Dyn_ arrow = D (arrow Dyn Dyn)
type Dyn = Dyn_ (->)
```
but got this error
```
Illegal Haskell construct encountered:
headed by: DArrowT
applied to: []typecheck
Q monad failuretypecheck
```
Is there a workaround.

**Comment by RyanGlScott** [comment]
Indeed, this is a pain point with the way `singletons-th` works. There is something of a workaround described in [this section of the README](https://github.com/goldfirere/singletons/tree/a169d3f6c0c8e962ea2983f60ed74e507bca9b2b#arrows-symbol-and-literals). For your use case, I imagine that would look something like this:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
module Dyn where

import Data.Singletons.TH
import Data.Singletons.TH.Options
import GHC.TypeLits.Singletons
import Language.Haskell.TH (Name)

-- Term-level
newtype Dyn  = D  ( Dyn ->  Dyn)
-- Type-level
newtype PDyn = PD (PDyn ~> PDyn)

$(let customPromote :: Name -> Name
      customPromote n
        | n == ''Dyn  = ''PDyn
        | n == 'D     = 'PD
        | otherwise   = promotedDataTypeOrConName defaultOptions n

      customDefun :: Name -> Int -> Name
      customDefun n sat = defunctionalizedName defaultOptions (customPromote n) sat in

  withOptions defaultOptions{ promotedDataTypeOrConName = customPromote
                            , defunctionalizedName      = customDefun
                            } $ do
    decs1 <- genSingletons [''Dyn]
    decs2 <- singletons [d|
               dynEx :: Dyn
               dynEx = D (\x -> x)
               
               -- Other definitions go here...
               |]
    return $ decs1 ++ decs2)
```

It's also worth noting that the `Illegal Haskell construct encountered` error message you encountered is really quite poor. I've filed a separate issue about this in #511.

**Comment by sweirich** [comment]
Thanks, Ryan. This worked.  I was able to create a type-level (untyped) lambda calculus evaluator by adding the following definitions:
```
               ($$) :: Dyn -> Dyn -> Dyn
               (D f) $$ x = f x

               -- lambda calculus
               data Nat = Zero | Succ Nat

               data Term = App Term Term | Lam Term | Var Nat

               eval :: Term -> [Dyn] -> Dyn
               eval (App t1 t2) e = eval t1 e $$ eval t2 e
               eval (Lam t) e = D $ \x -> eval t (x:e)
               eval (Var x) e = nlookup x e

               nlookup :: Nat -> [s] -> s
               nlookup Zero (x : env) = x
               nlookup (Succ n) (x : env) = nlookup n env
```

However, to do anything with these definitions, I had to assume the following extensionality axiom :-/  Do you think the axiom is sound?

```
extD :: forall f g. (forall x. Proxy x -> f $$ x :~: g $$ x) -> f :~: g
extD x = undefined 
```

**Comment by RyanGlScott** [comment]
> This worked.

Hooray!

> However, to do anything with these definitions, I had to assume the following extensionality axiom :-/ Do you think the axiom is sound?

Strictly speaking, no, as one can implement `unsafeCoerce` in terms of it:

```hs
-- DynEx1 and DynEx2 are defined in terms of defunctionalization symbols, and
-- each defunctionalization symbol is encoded as a strict data type. As a
-- result, GHC considers DynEx1 and DynEx2 to be apart, which means that a
-- pattern match on `DynEx1 :~: DynEx2` is considered unreachable. Therefore,
-- this is an exhaustive definition.
dynExesAren'tEqual :: DynEx1 :~: DynEx2 -> a
dynExesAren'tEqual x = case x of {}

-- On the other hand, DynEx1 and DynEx2 /are/ equal according to extD...
dynExesAreEqualAccordingToExtD :: DynEx1 :~: DynEx2
dynExesAreEqualAccordingToExtD = extD (\_ -> Refl)

-- ...which can lead to shenanigans.
unsafeCoerce :: a -> b
unsafeCoerce = dynExesAren'tEqual dynExesAreEqualAccordingToExtD
```

That being said, you probably won't do something like this. I've successfully used a definition like `extD` before in another project—see [`funExt` in the `ghc-software-foundations` repo](https://github.com/RyanGlScott/ghc-software-foundations/blob/ce7b8958e0aed4fb2c8611d71e7e0f1a2ef83222/logical-foundations/src/SF/LF/FunExt.hs#L10-L22). I was even bolder than you and defined `funExt` in terms of `unsafeCoerce`, not `undefined`. However, I had to take care to only apply `funExt` with a [special `apply` combinator](https://github.com/RyanGlScott/ghc-software-foundations/blob/ce7b8958e0aed4fb2c8611d71e7e0f1a2ef83222/logical-foundations/src/SF/LF/FunExt.hs#L58-L71) that was marked as `NOINLINE`. Otherwise, GHC's optimizer was liable to do strange things with it—see [GHC#16310](https://gitlab.haskell.org/ghc/ghc/-/issues/16310).

In short, while `extD` and the like are technically unsafe, they can be used safely in certain situations. Moreover, I don't know how to do better than `extD` in today's GHC, so I think `extD` is a reasonable thing to use.

**Comment by sweirich** [comment]
Yeah, that is what I was afraid of. And I can't see a workaround until we get type-level lambdas.

**Comment by RyanGlScott** [comment]
I think the main thrust of the issue (promoting data types with function types) has been addressed. Is there anything else to address here? Otherwise, I'll close this issue.

**Comment by sweirich** [comment]
Nope. Thanks for the discussion!
